File: python/src/coasty/dsl.py
"""Typed builders + validator for the Coasty workflow DSL (2026-06-01).

Step constructors: :func:`task`, :func:`assert_`, :func:`if_`, :func:`loop`,
:func:`parallel`, :func:`human_approval`, :func:`retry`, :func:`succeed`,
:func:`fail`. Condition constructors cover all 13 ops.

:func:`validate` enforces the documented limits client-side:

- <= 200 steps total (counting every nested step), <= 8 nesting levels,
- <= 16 parallel branches, ``retry.max_attempts`` in 1..20,
- no ``human_approval`` / ``succeed`` / ``fail`` inside a parallel branch,
- ``save_as`` not in ``{"inputs", "vars"}``, step ids match
  ``^[A-Za-z0-9_-]{1,64}$``.
"""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from typing import Any

from .types import CuaVersion, OnAwaitingHuman

Condition = dict[str, Any]
Step = dict[str, Any]
WorkflowDefinition = dict[str, Any]

DSL_VERSION = "2026-06-01"
MAX_TOTAL_STEPS = 200
MAX_NESTING_DEPTH = 8
MAX_PARALLEL_BRANCHES = 16
RETRY_ATTEMPTS_MIN = 1
RETRY_ATTEMPTS_MAX = 20
RESERVED_SAVE_AS = frozenset({"inputs", "vars"})
STEP_ID_RE = re.compile(r"^[A-Za-z0-9_-]{1,64}$")

STEP_TYPES = frozenset(
    {"task", "assert", "if", "loop", "parallel", "human_approval", "retry", "succeed", "fail"}
)
_COMPARISON_OPS = frozenset({"eq", "ne", "lt", "gt", "lte", "gte", "contains"})
_UNARY_VALUE_OPS = frozenset({"truthy", "falsy", "exists"})
_COMBINER_OPS = frozenset({"and", "or"})
CONDITION_OPS = _COMPARISON_OPS | _UNARY_VALUE_OPS | _COMBINER_OPS | {"not"}
_FORBIDDEN_IN_PARALLEL = frozenset({"human_approval", "succeed", "fail"})


class DSLValidationError(ValueError):
    """Raised by :func:`validate`; ``problems`` lists every violation found."""

    def __init__(self, problems: Sequence[str]) -> None:
        self.problems: list[str] = list(problems)
        joined = "\n  - ".join(self.problems)
        super().__init__(f"invalid workflow definition:\n  - {joined}")


def _drop_none(mapping: Mapping[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in mapping.items() if value is not None}


# ── Condition constructors (13 ops) ────────────────────────────────────────


def eq(left: Any, right: Any) -> Condition:
    return {"op": "eq", "left": left, "right": right}


def ne(left: Any, right: Any) -> Condition:
    return {"op": "ne", "left": left, "right": right}


def lt(left: Any, right: Any) -> Condition:
    return {"op": "lt", "left": left, "right": right}


def gt(left: Any, right: Any) -> Condition:
    return {"op": "gt", "left": left, "right": right}


def lte(left: Any, right: Any) -> Condition:
    return {"op": "lte", "left": left, "right": right}


def gte(left: Any, right: Any) -> Condition:
    return {"op": "gte", "left": left, "right": right}


def contains(left: Any, right: Any) -> Condition:
    return {"op": "contains", "left": left, "right": right}


def truthy(value: Any) -> Condition:
    return {"op": "truthy", "value": value}


def falsy(value: Any) -> Condition:
    return {"op": "falsy", "value": value}


def exists(value: Any) -> Condition:
    return {"op": "exists", "value": value}


def and_(*conditions: Condition) -> Condition:
    return {"op": "and", "conditions": list(conditions)}


def or_(*conditions: Condition) -> Condition:
    return {"op": "or", "conditions": list(conditions)}


def not_(condition: Condition) -> Condition:
    return {"op": "not", "condition": condition}


# ── Step constructors (9 types) ────────────────────────────────────────────


def task(
    step_id: str,
    task: str,
    *,
    machine_id: str | None = None,
    cua_version: CuaVersion | None = None,
    instructions: str | None = None,
    system_prompt: str | None = None,
    max_steps: int | None = None,
    save_as: str | None = None,
    on_awaiting_human: OnAwaitingHuman | None = None,
) -> Step:
    """A ``task`` step. Binds ``{status, passed, result, run_id, steps, error}``
    under ``save_as`` (and the step id)."""
    step: Step = {"id": step_id, "type": "task", "task": task}
    step.update(
        _drop_none(
            {
                "machine_id": machine_id,
                "cua_version": cua_version,
                "instructions": instructions,
                "system_prompt": system_prompt,
                "max_steps": max_steps,
                "save_as": save_as,
                "on_awaiting_human": on_awaiting_human,
            }
        )
    )
    return step


def assert_(step_id: str, condition: Condition, *, message: str | None = None) -> Step:
    """An ``assert`` step: fail the workflow unless ``condition`` holds."""
    step: Step = {"id": step_id, "type": "assert", "condition": condition}
    if message is not None:
        step["message"] = message
    return step


def if_(
    step_id: str,
    condition: Condition,
    then: Sequence[Step],
    *,
    else_: Sequence[Step] | None = None,
) -> Step:
    """An ``if`` step branching on a structured condition."""
    step: Step = {"id": step_id, "type": "if", "condition": condition, "then": list(then)}
    if else_ is not None:
        step["else"] = list(else_)
    return step


def loop(
    step_id: str,
    body: Sequence[Step],
    *,
    count: int | None = None,
    while_: Condition | None = None,
    max_iterations: int | None = None,
) -> Step:
    """A ``loop`` step. Exactly one of ``count`` / ``while_`` is required."""
    if (count is None) == (while_ is None):
        raise ValueError("loop() requires exactly one of count= or while_=")
    step: Step = {"id": step_id, "type": "loop", "body": list(body)}
    if count is not None:
        step["count"] = count
    if while_ is not None:
        step["while"] = while_
    if max_iterations is not None:
        step["max_iterations"] = max_iterations
    return step


def parallel(step_id: str, branches: Sequence[Sequence[Step]]) -> Step:
    """A ``parallel`` step running independent branches concurrently."""
    return {"id": step_id, "type": "parallel", "branches": [list(b) for b in branches]}


def human_approval(
    step_id: str,
    *,
    message: str | None = None,
    timeout_seconds: int | None = None,
) -> Step:
    """A ``human_approval`` step: pause until approved/rejected."""
    step: Step = {"id": step_id, "type": "human_approval"}
    step.update(_drop_none({"message": message, "timeout_seconds": timeout_seconds}))
    return step


def retry(step_id: str, body: Sequence[Step], *, max_attempts: int) -> Step:
    """A ``retry`` step: re-run ``body`` up to ``max_attempts`` times (1-20)."""
    return {"id": step_id, "type": "retry", "body": list(body), "max_attempts": max_attempts}


def succeed(step_id: str, *, output: Mapping[str, Any] | None = None) -> Step:
    """A ``succeed`` step finishing the workflow with an optional output."""
    step: Step = {"id": step_id, "type": "succeed"}
    if output is not None:
        step["output"] = dict(output)
    return step


def fail(step_id: str, *, message: str | None = None) -> Step:
    """A ``fail`` step finishing the workflow as failed."""
    step: Step = {"id": step_id, "type": "fail"}
    if message is not None:
        step["message"] = message
    return step


def definition(
    steps: Sequence[Step],
    *,
    output: Mapping[str, Any] | None = None,
) -> WorkflowDefinition:
    """Wrap steps (and an optional top-level ``output``) into a definition."""
    result: WorkflowDefinition = {"steps": list(steps)}
    if output is not None:
        result["output"] = dict(output)
    return result


# ── Validation ─────────────────────────────────────────────────────────────


def _validate_condition(condition: object, where: str, problems: list[str]) -> None:
    if not isinstance(condition, Mapping):
        problems.append(f"{where}: condition must be an object, got {type(condition).__name__}")
        return
    op = condition.get("op")
    if not isinstance(op, str) or op not in CONDITION_OPS:
        problems.append(f"{where}: unknown condition op {op!r}")
        return
    if op in _COMPARISON_OPS:
        for key in ("left", "right"):
            if key not in condition:
                problems.append(f"{where}: condition op {op!r} requires {key!r}")
    elif op in _UNARY_VALUE_OPS:
        if "value" not in condition:
            problems.append(f"{where}: condition op {op!r} requires 'value'")
    elif op in _COMBINER_OPS:
        nested = condition.get("conditions")
        if not isinstance(nested, list) or not nested:
            problems.append(f"{where}: condition op {op!r} requires a non-empty 'conditions' list")
        else:
            for index, child in enumerate(nested):
                _validate_condition(child, f"{where}.conditions[{index}]", problems)
    else:  # not
        if "condition" not in condition:
            problems.append(f"{where}: condition op 'not' requires 'condition'")
        else:
            _validate_condition(condition["condition"], f"{where}.condition", problems)


def _validate_steps(
    steps: object,
    *,
    where: str,
    depth: int,
    inside_parallel: bool,
    problems: list[str],
) -> int:
    """Validate a step list; returns the number of steps counted (recursively)."""
    if not isinstance(steps, list):
        problems.append(f"{where}: steps must be a list, got {type(steps).__name__}")
        return 0
    if depth > MAX_NESTING_DEPTH:
        problems.append(
            f"{where}: nesting depth {depth} exceeds the maximum of {MAX_NESTING_DEPTH}"
        )
        return len(steps)

    count = 0
    for index, step in enumerate(steps):
        label = f"{where}[{index}]"
        count += 1
        if not isinstance(step, Mapping):
            problems.append(f"{label}: step must be an object, got {type(step).__name__}")
            continue

        step_id = step.get("id")
        if not isinstance(step_id, str) or not STEP_ID_RE.fullmatch(step_id):
            problems.append(f"{label}: step id {step_id!r} must match ^[A-Za-z0-9_-]{{1,64}}$")
        else:
            label = f"{where}[{index}] (id={step_id})"

        step_type = step.get("type")
        if not isinstance(step_type, str) or step_type not in STEP_TYPES:
            problems.append(f"{label}: unknown step type {step_type!r}")
            continue

        if inside_parallel and step_type in _FORBIDDEN_IN_PARALLEL:
            problems.append(
                f"{label}: {step_type!r} steps are not allowed inside a parallel branch"
            )

        if step_type == "task":
            task_text = step.get("task")
            if not isinstance(task_text, str) or not task_text:
                problems.append(f"{label}: task steps require a non-empty 'task' string")
            save_as = step.get("save_as")
            if save_as is not None:
                if not isinstance(save_as, str):
                    problems.append(f"{label}: save_as must be a string")
                elif save_as in RESERVED_SAVE_AS:
                    problems.append(f"{label}: save_as {save_as!r} is reserved (inputs, vars)")
        elif step_type == "assert":
            if "condition" not in step:
                problems.append(f"{label}: assert steps require a 'condition'")
            else:
                _validate_condition(step["condition"], label, problems)
        elif step_type == "if":
            if "condition" not in step:
                problems.append(f"{label}: if steps require a 'condition'")
            else:
                _validate_condition(step["condition"], label, problems)
            if "then" not in step:
                problems.append(f"{label}: if steps require a 'then' branch")
            else:
                count += _validate_steps(
                    step["then"],
                    where=f"{label}.then",
                    depth=depth + 1,
                    inside_parallel=inside_parallel,
                    problems=problems,
                )
            if "else" in step:
                count += _validate_steps(
                    step["else"],
                    where=f"{label}.else",
                    depth=depth + 1,
                    inside_parallel=inside_parallel,
                    problems=problems,
                )
        elif step_type == "loop":
            has_count = "count" in step
            has_while = "while" in step
            if has_count == has_while:
                problems.append(f"{label}: loop steps require exactly one of 'count' or 'while'")
            if has_count and (not isinstance(step["count"], int) or step["count"] < 1):
                problems.append(f"{label}: loop count must be an integer >= 1")
            if has_while:
                _validate_condition(step["while"], f"{label}.while", problems)
            max_iterations = step.get("max_iterations")
            if max_iterations is not None and (
                not isinstance(max_iterations, int) or max_iterations < 1
            ):
                problems.append(f"{label}: max_iterations must be an integer >= 1")
            if "body" not in step:
                problems.append(f"{label}: loop steps require a 'body'")
            else:
                count += _validate_steps(
                    step["body"],
                    where=f"{label}.body",
                    depth=depth + 1,
                    inside_parallel=inside_parallel,
                    problems=problems,
                )
        elif step_type == "parallel":
            branches = step.get("branches")
            if not isinstance(branches, list) or not branches:
                problems.append(f"{label}: parallel steps require a non-empty 'branches' list")
            else:
                if len(branches) > MAX_PARALLEL_BRANCHES:
                    problems.append(
                        f"{label}: {len(branches)} parallel branches exceed the "
                        f"maximum of {MAX_PARALLEL_BRANCHES}"
                    )
                for branch_index, branch in enumerate(branches):
                    count += _validate_steps(
                        branch,
                        where=f"{label}.branches[{branch_index}]",
                        depth=depth + 1,
                        inside_parallel=True,
                        problems=problems,
                    )
        elif step_type == "retry":
            max_attempts = step.get("max_attempts")
            if (
                not isinstance(max_attempts, int)
                or isinstance(max_attempts, bool)
                or not RETRY_ATTEMPTS_MIN <= max_attempts <= RETRY_ATTEMPTS_MAX
            ):
                problems.append(
                    f"{label}: retry max_attempts must be an integer in "
                    f"{RETRY_ATTEMPTS_MIN}..{RETRY_ATTEMPTS_MAX}"
                )
            if "body" not in step:
                problems.append(f"{label}: retry steps require a 'body'")
            else:
                count += _validate_steps(
                    step["body"],
                    where=f"{label}.body",
                    depth=depth + 1,
                    inside_parallel=inside_parallel,
                    problems=problems,
                )
        # human_approval / succeed / fail need no extra field checks
    return count


def validate(definition: Mapping[str, Any]) -> None:
    """Validate a workflow definition against the documented limits.

    Raises :class:`DSLValidationError` (with a ``problems`` list) on any
    violation; returns ``None`` when the definition is structurally valid.
    """
    problems: list[str] = []
    if not isinstance(definition, Mapping):
        raise DSLValidationError([f"definition must be an object, got {type(definition).__name__}"])
    if "steps" not in definition:
        problems.append("definition requires a 'steps' list")
        raise DSLValidationError(problems)

    total = _validate_steps(
        definition["steps"],
        where="steps",
        depth=1,
        inside_parallel=False,
        problems=problems,
    )
    if total > MAX_TOTAL_STEPS:
        problems.append(
            f"definition has {total} steps (counting nested); the maximum is {MAX_TOTAL_STEPS}"
        )
    if problems:
        raise DSLValidationError(problems)

File: typescript/src/coasty/dsl.ts
/**
 * Typed builders + structural validation for the Coasty Workflow DSL
 * (dsl_version 2026-06-01).
 *
 * 9 step types: task, assert, if, loop, parallel, human_approval, retry,
 * succeed, fail. 13 condition ops. `CONDITION_OPS` is a readonly tuple so the
 * type IS the contract.
 */
import { type CuaVersion, type OnAwaitingHuman } from './types.js';

// ---------------------------------------------------------------------------
// Conditions (13 ops — injection-safe, no free-text eval)
// ---------------------------------------------------------------------------

export const COMPARISON_OPS = ['eq', 'ne', 'lt', 'gt', 'lte', 'gte', 'contains'] as const;
export const VALUE_OPS = ['truthy', 'falsy', 'exists'] as const;
export const COMPOSITE_OPS = ['and', 'or'] as const;
export const NEGATION_OPS = ['not'] as const;

/** All 13 documented condition ops. This tuple IS the contract. */
export const CONDITION_OPS = [
  ...COMPARISON_OPS,
  ...VALUE_OPS,
  ...COMPOSITE_OPS,
  ...NEGATION_OPS,
] as const;

export type ComparisonOp = (typeof COMPARISON_OPS)[number];
export type ValueOp = (typeof VALUE_OPS)[number];
export type CompositeOp = (typeof COMPOSITE_OPS)[number];
export type NegationOp = (typeof NEGATION_OPS)[number];
export type ConditionOp = (typeof CONDITION_OPS)[number];

/** Operands may be literals or `{{path}}` template strings. */
export type TemplateValue = string | number | boolean | null;

export interface ComparisonCondition {
  op: ComparisonOp;
  left: TemplateValue;
  right: TemplateValue;
}

export interface ValueCondition {
  op: ValueOp;
  value: TemplateValue;
}

export interface CompositeCondition {
  op: CompositeOp;
  conditions: Condition[];
}

export interface NotCondition {
  op: NegationOp;
  condition: Condition;
}

export type Condition = ComparisonCondition | ValueCondition | CompositeCondition | NotCondition;

// ---------------------------------------------------------------------------
// Steps (9 types)
// ---------------------------------------------------------------------------

export const STEP_TYPES = [
  'task',
  'assert',
  'if',
  'loop',
  'parallel',
  'human_approval',
  'retry',
  'succeed',
  'fail',
] as const;
export type StepType = (typeof STEP_TYPES)[number];

/** Step ids must match this pattern. */
export const STEP_ID_PATTERN = /^[A-Za-z0-9_-]{1,64}$/;
/** `save_as` may not shadow these namespaces. */
export const RESERVED_SAVE_AS = ['inputs', 'vars'] as const;

export const MAX_TOTAL_STEPS = 200;
export const MAX_NESTING_DEPTH = 8;
export const MAX_PARALLEL_BRANCHES = 16;
export const MIN_RETRY_ATTEMPTS = 1;
export const MAX_RETRY_ATTEMPTS = 20;

export interface TaskStep {
  id: string;
  type: 'task';
  /** Supports `{{inputs.x}}` / `{{vars.y}}` / `{{stepId.field}}` templating. */
  task: string;
  machine_id?: string;
  cua_version?: CuaVersion;
  instructions?: string;
  system_prompt?: string;
  max_steps?: number;
  /** Binds `{status, passed, result, run_id, steps, error}` under this name. */
  save_as?: string;
  on_awaiting_human?: OnAwaitingHuman;
}

export interface AssertStep {
  id: string;
  type: 'assert';
  condition: Condition;
  message?: string;
}

export interface IfStep {
  id: string;
  type: 'if';
  condition: Condition;
  then: WorkflowStep[];
  else?: WorkflowStep[];
}

export interface LoopStep {
  id: string;
  type: 'loop';
  /** Exactly one of `count` | `while` must be set. */
  count?: number;
  while?: Condition;
  body: WorkflowStep[];
  max_iterations?: number;
}

export interface ParallelStep {
  id: string;
  type: 'parallel';
  /** <= 16 branches; no human_approval/succeed/fail anywhere inside. */
  branches: WorkflowStep[][];
}

export interface HumanApprovalStep {
  id: string;
  type: 'human_approval';
  message?: string;
  timeout_seconds?: number;
}

export interface RetryStep {
  id: string;
  type: 'retry';
  body: WorkflowStep[];
  /** 1-20. */
  max_attempts: number;
}

export interface SucceedStep {
  id: string;
  type: 'succeed';
  output?: Record<string, unknown>;
}

export interface FailStep {
  id: string;
  type: 'fail';
  message?: string;
}

export type WorkflowStep =
  | TaskStep
  | AssertStep
  | IfStep
  | LoopStep
  | ParallelStep
  | HumanApprovalStep
  | RetryStep
  | SucceedStep
  | FailStep;

export interface WorkflowDefinition {
  steps: WorkflowStep[];
  output?: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Step builders
// ---------------------------------------------------------------------------

export function definition(
  steps: WorkflowStep[],
  output?: Record<string, unknown>,
): WorkflowDefinition {
  return output === undefined ? { steps } : { steps, output };
}

export function task(
  id: string,
  taskText: string,
  options: Omit<TaskStep, 'id' | 'type' | 'task'> = {},
): TaskStep {
  return { id, type: 'task', task: taskText, ...options };
}

export function assertStep(id: string, condition: Condition, message?: string): AssertStep {
  return message === undefined
    ? { id, type: 'assert', condition }
    : { id, type: 'assert', condition, message };
}

export function ifStep(
  id: string,
  condition: Condition,
  thenSteps: WorkflowStep[],
  elseSteps?: WorkflowStep[],
): IfStep {
  return elseSteps === undefined
    ? { id, type: 'if', condition, then: thenSteps }
    : { id, type: 'if', condition, then: thenSteps, else: elseSteps };
}

export function loopCount(
  id: string,
  count: number,
  body: WorkflowStep[],
  maxIterations?: number,
): LoopStep {
  return maxIterations === undefined
    ? { id, type: 'loop', count, body }
    : { id, type: 'loop', count, body, max_iterations: maxIterations };
}

export function loopWhile(
  id: string,
  condition: Condition,
  body: WorkflowStep[],
  maxIterations?: number,
): LoopStep {
  return maxIterations === undefined
    ? { id, type: 'loop', while: condition, body }
    : { id, type: 'loop', while: condition, body, max_iterations: maxIterations };
}

export function parallel(id: string, branches: WorkflowStep[][]): ParallelStep {
  return { id, type: 'parallel', branches };
}

export function humanApproval(
  id: string,
  options: { message?: string; timeoutSeconds?: number } = {},
): HumanApprovalStep {
  const step: HumanApprovalStep = { id, type: 'human_approval' };
  if (options.message !== undefined) step.message = options.message;
  if (options.timeoutSeconds !== undefined) step.timeout_seconds = options.timeoutSeconds;
  return step;
}

export function retryStep(id: string, body: WorkflowStep[], maxAttempts: number): RetryStep {
  return { id, type: 'retry', body, max_attempts: maxAttempts };
}

export function succeed(id: string, output?: Record<string, unknown>): SucceedStep {
  return output === undefined ? { id, type: 'succeed' } : { id, type: 'succeed', output };
}

export function fail(id: string, message?: string): FailStep {
  return message === undefined ? { id, type: 'fail' } : { id, type: 'fail', message };
}

// ---------------------------------------------------------------------------
// Condition builders
// ---------------------------------------------------------------------------

function comparison(
  op: ComparisonOp,
  left: TemplateValue,
  right: TemplateValue,
): ComparisonCondition {
  return { op, left, right };
}

export const eq = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('eq', left, right);
export const ne = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('ne', left, right);
export const lt = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('lt', left, right);
export const gt = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('gt', left, right);
export const lte = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('lte', left, right);
export const gte = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('gte', left, right);
export const contains = (left: TemplateValue, right: TemplateValue): ComparisonCondition =>
  comparison('contains', left, right);

export const truthy = (value: TemplateValue): ValueCondition => ({ op: 'truthy', value });
export const falsy = (value: TemplateValue): ValueCondition => ({ op: 'falsy', value });
export const exists = (value: TemplateValue): ValueCondition => ({ op: 'exists', value });

export const and = (...conditions: Condition[]): CompositeCondition => ({ op: 'and', conditions });
export const or = (...conditions: Condition[]): CompositeCondition => ({ op: 'or', conditions });
export const not = (condition: Condition): NotCondition => ({ op: 'not', condition });

// ---------------------------------------------------------------------------
// Validation (documented limits, enforced at create / ad-hoc time)
// ---------------------------------------------------------------------------

export interface DslIssue {
  /** JSON-path-ish location, e.g. `steps[0].branches[1][0]`. */
  path: string;
  message: string;
}

/** Thrown by {@link assertValidDefinition} when a definition violates the DSL limits. */
export class WorkflowDslError extends Error {
  readonly issues: readonly DslIssue[];

  constructor(issues: readonly DslIssue[]) {
    const lines = issues.map((issue) => `  - ${issue.path}: ${issue.message}`).join('\n');
    super(`Invalid workflow definition (${String(issues.length)} issue(s)):\n${lines}`);
    this.name = 'WorkflowDslError';
    this.issues = issues;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

interface WalkState {
  issues: DslIssue[];
  totalSteps: number;
  insideParallel: boolean;
}

const PARALLEL_FORBIDDEN: ReadonlySet<string> = new Set(['human_approval', 'succeed', 'fail']);

function validateCondition(condition: unknown, path: string, issues: DslIssue[]): void {
  if (!isRecord(condition)) {
    issues.push({ path, message: 'condition must be an object with an "op" field' });
    return;
  }
  const op = condition.op;
  if (typeof op !== 'string' || !(CONDITION_OPS as readonly string[]).includes(op)) {
    issues.push({
      path,
      message: `unknown condition op ${JSON.stringify(op)} (expected one of: ${CONDITION_OPS.join(', ')})`,
    });
    return;
  }
  if ((COMPARISON_OPS as readonly string[]).includes(op)) {
    if (!('left' in condition) || !('right' in condition)) {
      issues.push({ path, message: `op "${op}" requires "left" and "right"` });
    }
  } else if ((VALUE_OPS as readonly string[]).includes(op)) {
    if (!('value' in condition)) {
      issues.push({ path, message: `op "${op}" requires "value"` });
    }
  } else if ((COMPOSITE_OPS as readonly string[]).includes(op)) {
    const conditions = condition.conditions;
    if (!Array.isArray(conditions) || conditions.length === 0) {
      issues.push({ path, message: `op "${op}" requires a non-empty "conditions" array` });
    } else {
      conditions.forEach((nested, index) => {
        validateCondition(nested, `${path}.conditions[${String(index)}]`, issues);
      });
    }
  } else {
    // 'not'
    if (!('condition' in condition)) {
      issues.push({ path, message: 'op "not" requires "condition"' });
    } else {
      validateCondition(condition.condition, `${path}.condition`, issues);
    }
  }
}

function validateSteps(steps: unknown, path: string, depth: number, state: WalkState): void {
  if (!Array.isArray(steps)) {
    state.issues.push({ path, message: 'expected an array of steps' });
    return;
  }
  if (depth > MAX_NESTING_DEPTH) {
    state.issues.push({
      path,
      message: `steps nest deeper than the maximum of ${String(MAX_NESTING_DEPTH)} levels`,
    });
    return;
  }
  steps.forEach((step, index) => {
    validateStep(step, `${path}[${String(index)}]`, depth, state);
  });
}

function validateStep(step: unknown, path: string, depth: number, state: WalkState): void {
  state.totalSteps += 1;
  if (state.totalSteps === MAX_TOTAL_STEPS + 1) {
    state.issues.push({
      path,
      message: `definition exceeds the maximum of ${String(MAX_TOTAL_STEPS)} total steps (counting nested steps)`,
    });
  }
  if (!isRecord(step)) {
    state.issues.push({ path, message: 'step must be an object' });
    return;
  }

  const id = step.id;
  if (typeof id !== 'string' || !STEP_ID_PATTERN.test(id)) {
    state.issues.push({
      path,
      message: `step id ${JSON.stringify(id)} must match ${STEP_ID_PATTERN.source}`,
    });
  }

  const type = step.type;
  if (typeof type !== 'string' || !(STEP_TYPES as readonly string[]).includes(type)) {
    state.issues.push({
      path,
      message: `unknown step type ${JSON.stringify(type)} (expected one of: ${STEP_TYPES.join(', ')})`,
    });
    return;
  }

  if (state.insideParallel && PARALLEL_FORBIDDEN.has(type)) {
    state.issues.push({
      path,
      message: `step type "${type}" is not allowed inside a parallel branch`,
    });
  }

  switch (type) {
    case 'task': {
      if (typeof step.task !== 'string' || step.task.length === 0) {
        state.issues.push({ path, message: 'task step requires a non-empty "task" string' });
      }
      const saveAs = step.save_as;
      if (saveAs !== undefined) {
        if (typeof saveAs !== 'string') {
          state.issues.push({ path, message: '"save_as" must be a string' });
        } else if ((RESERVED_SAVE_AS as readonly string[]).includes(saveAs)) {
          state.issues.push({
            path,
            message: `"save_as" must not be a reserved namespace (${RESERVED_SAVE_AS.join(', ')})`,
          });
        }
      }
      break;
    }
    case 'assert': {
      validateCondition(step.condition, `${path}.condition`, state.issues);
      break;
    }
    case 'if': {
      validateCondition(step.condition, `${path}.condition`, state.issues);
      validateSteps(step.then, `${path}.then`, depth + 1, state);
      if (step.else !== undefined) validateSteps(step.else, `${path}.else`, depth + 1, state);
      break;
    }
    case 'loop': {
      const hasCount = step.count !== undefined;
      const hasWhile = step.while !== undefined;
      if (hasCount === hasWhile) {
        state.issues.push({
          path,
          message: 'loop step requires exactly one of "count" or "while"',
        });
      }
      if (
        hasCount &&
        (typeof step.count !== 'number' || !Number.isInteger(step.count) || step.count < 1)
      ) {
        state.issues.push({ path, message: '"count" must be a positive integer' });
      }
      if (hasWhile) validateCondition(step.while, `${path}.while`, state.issues);
      validateSteps(step.body, `${path}.body`, depth + 1, state);
      break;
    }
    case 'parallel': {
      const branches = step.branches;
      if (!Array.isArray(branches)) {
        state.issues.push({ path, message: 'parallel step requires a "branches" array of arrays' });
        break;
      }
      if (branches.length > MAX_PARALLEL_BRANCHES) {
        state.issues.push({
          path,
          message: `parallel step has ${String(branches.length)} branches (maximum is ${String(MAX_PARALLEL_BRANCHES)})`,
        });
      }
      const wasInsideParallel = state.insideParallel;
      state.insideParallel = true;
      branches.forEach((branch, index) => {
        validateSteps(branch, `${path}.branches[${String(index)}]`, depth + 1, state);
      });
      state.insideParallel = wasInsideParallel;
      break;
    }
    case 'retry': {
      const attempts = step.max_attempts;
      if (
        typeof attempts !== 'number' ||
        !Number.isInteger(attempts) ||
        attempts < MIN_RETRY_ATTEMPTS ||
        attempts > MAX_RETRY_ATTEMPTS
      ) {
        state.issues.push({
          path,
          message: `"max_attempts" must be an integer between ${String(MIN_RETRY_ATTEMPTS)} and ${String(MAX_RETRY_ATTEMPTS)}`,
        });
      }
      validateSteps(step.body, `${path}.body`, depth + 1, state);
      break;
    }
    case 'human_approval':
    case 'succeed':
    case 'fail':
      break;
  }
}

/**
 * Structurally validate a workflow definition against the documented limits:
 * <= 200 total steps (nested counted), <= 8 nesting levels, <= 16 parallel
 * branches, retry attempts 1-20, no human_approval/succeed/fail inside
 * parallel, reserved save_as names, and the step-id regex.
 *
 * Returns a list of issues (empty = valid). Defensive: accepts arbitrary
 * runtime data without throwing.
 */
export function validateDefinition(def: WorkflowDefinition): DslIssue[] {
  const state: WalkState = { issues: [], totalSteps: 0, insideParallel: false };
  const raw: unknown = def;
  if (!isRecord(raw)) {
    return [{ path: '$', message: 'definition must be an object with a "steps" array' }];
  }
  validateSteps(raw.steps, 'steps', 1, state);
  return state.issues;
}

/** Like {@link validateDefinition} but throws {@link WorkflowDslError} on any issue. */
export function assertValidDefinition(def: WorkflowDefinition): void {
  const issues = validateDefinition(def);
  if (issues.length > 0) throw new WorkflowDslError(issues);
}
